### PR TITLE
DM-35878: escalate alarms to OpsGenie

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

File: python/lsst/ts/watcher/alarm.py [hunk 1]
```diff
@@ -150,21 +157,57 @@ def configure(
...
+            or escalation_delay = 0 and escalation_responders not empty.
+        """
+        if escalation_delay < 0:
+            raise ValueError(f"{escalation_delay=} must be ≥ 0")
+        if (escalation_delay == 0) != (len(escalation_responders) == 0):
+            raise ValueError(
+                f"{escalation_delay=} must be > 0 if and only if"
+                f"{escalation_responders=} is not empty"
+            )
+        self.escalation_responders = list(escalation_responders)
```
Comment (wvreeven) [inline python/lsst/ts/watcher/alarm.py:194]: Why is an explicit cast to `list` needed here if the doc string for `escalation_responders` already indicates that it is a list?

File: python/lsst/ts/watcher/mock_opsgenie.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+def get_request_id() -> str:
+    """Return a new unique request ID"""
+    return str(uuid.uuid4())
+
+
+def make_processing_reply(request_id: str) -> web.Reponse:
+    """Make a standard "processing" reply.
+
+    Parameter
```
Comment (wvreeven) [inline python/lsst/ts/watcher/mock_opsgenie.py:41]: `Parameter` -> `Parameters`

File: python/lsst/ts/watcher/mock_opsgenie.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            raise web.HTTPForbidden(
+                text="Authorization header must be of the form 'GenieKey <key>'"
+            )
+        if name != "GenieKey":
+            raise web.HTTPForbidden(
+                text="Authorization header must start with GenieKey"
+            )
+        if value != self.escalation_key:
+            raise web.HTTPForbidden(text="Authorization key does not match")
+        return None
```
Comment (wvreeven) [inline python/lsst/ts/watcher/mock_opsgenie.py:170]: This line may be removed, right?

File: python/lsst/ts/watcher/mock_opsgenie.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+    async def close_alert(self, request) -> web.json_response:
+        """Request handler to set the status of an alert to "closed"."""
```
Comment (wvreeven) [inline python/lsst/ts/watcher/mock_opsgenie.py:173]: Can you please complete the doc string for this function and describe the `request` parameters as well as the return value. Also for `create_alert` and `delete_alert` please.

File: python/lsst/ts/watcher/watcher_csc.py [hunk 5]
```diff
@@ -103,24 +124,151 @@ def _enable_or_disable_model(self):
...
+                else:
+                    read_text = await response.text()
+                    alarm.escalated_id = f"Failed: {read_text}"
+                    self.log.warning(f"Could not escalate alarm {alarm}: {read_text}")
+        except Exception as e:
+            errmsg = f"Could not reach OpsGenie: {e!r}"
+            alarm.escalated_id = f"Failed: {errmsg}"
+            self.log.warning(f"Could not escalate alarm {alarm}: {errmsg}")
+
+    async def deescalte_alarm(self, alarm):
```
Comment (wvreeven) [inline python/lsst/ts/watcher/watcher_csc.py:182]: `deescalte_alarm` -> `deescalate_alarm`

**Review by wvreeven — APPROVED** [review]
In general this looks great. I particularly like the mock OpsGenie. I did find a few typos and other things I commented on. I leave it to you to process or ignore them.

File: python/lsst/ts/watcher/alarm.py (hunk 1 above)
Comment (r-owen) [reply python/lsst/ts/watcher/alarm.py:194]: I don't try for fully correct types in doc strings (only in formal Python type annotations). list is the shortcut I always for any generic ordered sequence.

File: python/lsst/ts/watcher/mock_opsgenie.py (hunk 4 above)
Comment (r-owen) [reply python/lsst/ts/watcher/mock_opsgenie.py:173]: Yes. I'll add the missing type annotation, as well. I don't know how to make the  description very informative, but it's still better to have it.

**Pull request merged**